### PR TITLE
Lazy fetch definitions

### PR DESCRIPTION
**Pull request by smancke**

### CODE REVIEW

File: composition/content_fetcher.go [hunk 1]
```diff
@@ -115,9 +129,24 @@ func (fetcher *ContentFetcher) AddFetchJob(d *FetchDefinition) {
...
+			}
+			for dependencyName, params := range fetchResult.Content.Dependencies() {
+				_, exist := fetcher.r.sheduledFetchDefinitionNames[dependencyName]
+				if !exist {
+					lazyFd, exist, err := fetcher.lazyFdFactory(dependencyName, params)
+					if err != nil {
+						logging.Logger.WithError(err).
+							WithField("dependencyName", dependencyName).
+							WithField("params", params).
+							Errorf("failed fetching dependency %v", dependencyName)
```
Comment (domano) [inline composition/content_fetcher.go]: The error message is telling the wrong thing. The dependency ist not being fetched here. The error should be something like "failed creating fetchdefinition for dependency..."

File: composition/content_fetcher.go [hunk 2]
```diff
@@ -115,9 +129,24 @@ func (fetcher *ContentFetcher) AddFetchJob(d *FetchDefinition) {
...
 		if fetchResult.Err == nil {
 			fetcher.addMeta(fetchResult.Content.Meta())
-			for _, dependency := range fetchResult.Content.RequiredContent() {
-				if dependency.IsFetchable() {
-					fetcher.AddFetchJob(dependency)
+			for _, fetch := range fetchResult.Content.RequiredContent() {
+				fetcher.AddFetchJob(fetch)
+			}
+			for dependencyName, params := range fetchResult.Content.Dependencies() {
+				_, exist := fetcher.r.sheduledFetchDefinitionNames[dependencyName]
```
Comment (domano) [inline composition/content_fetcher.go]: exist is a verb (=> method name), i would use "existing" or something else.

File: composition/content_fetcher.go [hunk 3]
```diff
@@ -33,18 +27,25 @@ func (fr FetchResults) Less(i, j int) bool {
 	return fr[i].Def.Priority < fr[j].Def.Priority
 }
 
+// FetchDefinitionFactory should return a fetch definition for the given name and parameters.
+// This factory method can be used to supply lazy loaded fetch jobs.
+// The FetchDefinition returned has to have the same name as the supplied name parameter.
+type FetchDefinitionFactory func(name string, params Params) (fd *FetchDefinition, exist bool, err error)
```
Comment (domano) [inline composition/content_fetcher.go]: The meaning of the "exist" return value is unclear. Maybe add a comment describing its meaning. (Especially as there is another variable named "exist" in the AddFetchJob method)

**Comment by smancke** [comment]
Hey @domano, I extracted the block and renamed the exist variables.
Please check and push the merge button, if everything is fine, now.

**Pull request merged**